### PR TITLE
FI-930: Support FHIRPath for generating searches

### PR DESCRIPTION
**Pull request by jason-crowley**
[FI-930](https://oncprojectracking.healthit.gov/support/browse/FI-930): SearchParameters that have complicated FHIRPath expressions (e.g. use descendants() or |) cannot be generically processed by Inferno.  Inferno uses the FHIRPath in order to determine the value of the searchParameter to be used in the search and validate that the results returned conform to the value.

This PR creates a `FHIRPathEvaluator` class that posts elements to the FHIRPath endpoint. It also adds two new config options, `fhirpath_evaluator` and `external_fhirpath_evaluator_url`, which are analogous to the `resource_validator` and `external_resource_validator_url` config options. When `fhirpath_evaluator` is set to `external`, POST requests will be made to the FHIRPath endpoint in order to resolve FHIRPath expressions against a given complex datatype.

The goal was to substitute out the logic that was previously `resolve_path` and `resolve_element_from_path` with a more complete FHIRPath solution. This allows us to remove some of the FHIRPath patching (e.g. removing `.as()` and truncating the path) that was done in the sequences. However, one patch that has not been removed is `.where(resolve() ...)` because the external FHIRPath evaluator does not know how to resolve references of the form `[Resource]/[id]`.

~NOTE: This PR requires running [wrapper/FI-931-add-fhirpath-endpoint](https://github.com/inferno-community/fhir-validator-wrapper/pull/19) locally to work.~ UPDATE: this is no longer true; you will just have to re-pull the `master` branch of `f-v-w` and rebuild.

**Submitter:**
- [X] This pull request describes why these changes were made
- [X] Internal ticket for this PR: https://oncprojectracking.healthit.gov/support/browse/FI-930
- [X] Internal ticket links to this PR
- [X] Internal ticket is properly labeled (Community/Program)
- [X] Internal ticket has a justification for its Community/Program label
- [X] Code diff has been reviewed for extraneous/missing code
- [X] Tests are included and test edge cases
- [X] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: lib/app/utils/fhirpath_evaluator.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    # Evaluates the given FHIRPath expression against the given elements by posting each element
+    # to the FHIRPath wrapper.
+    # @param elements [Array]
+    # @param path [String]
+    def evaluate(elements, path)
+      elements = Array.wrap(elements)
+      return elements if path.blank?
+
+      types = elements.map { |e| e.class.to_s.split('::').last }
```
Comment (Jammjammjamm) [inline lib/app/utils/fhirpath_evaluator.rb:33]:
```suggestion
      types = elements.map { |e| e.class.name.demodulize }
```
ActiveSupport to the rescue once again!

File: lib/app/utils/fhirpath_evaluator.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      raise ArgumentError, 'FHIRPath URL is unset' if fhirpath_url.blank?
+
+      @fhirpath_url = fhirpath_url
+    end
+
+    # Evaluates the given FHIRPath expression against the given elements by posting each element
+    # to the FHIRPath wrapper.
+    # @param elements [Array]
+    # @param path [String]
+    def evaluate(elements, path)
```
Comment (Jammjammjamm) [inline lib/app/utils/fhirpath_evaluator.rb:29]: Do we have a need to evaluate fhirpath expressions against arbitrary elements rather than against entire FHIR resources?

File: lib/app/utils/fhirpath_evaluator.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    # @param selected_evaluator [String] either 'internal' or 'external'
+    # @param external_evaluator_url [String] the base url for the external evaluator
+    def self.new_evaluator(selected_evaluator, external_evaluator_url)
+      return nil unless ENV['RACK_ENV'] != 'test' && selected_evaluator == 'external'
+
+      Inferno::FHIRPathEvaluator.new(external_evaluator_url)
+    end
+
+    # @param fhirpath_url [String] the base url for the FHIRPath /evaluate endpoint
+    def initialize(fhirpath_url)
```
Comment (Jammjammjamm) [inline lib/app/utils/fhirpath_evaluator.rb:19]: Looking over this branch is exposing some issues that are also present in the validator implementation. This class (as well as both of the validator classes) has no non-global state, so it could just be a module instead. This isn't necessarily something that needs fixing, just something that jumped out at me.

File: lib/app/utils/fhirpath_evaluator.rb (hunk 2 above)
Comment (jason-crowley) [reply lib/app/utils/fhirpath_evaluator.rb:29]: Yes, we need it to evaluate paths against CodeableConcepts and the Provenance.agent BackboneElement.

File: lib/app/sequence_base.rb [hunk 4]
```diff
@@ -743,6 +743,8 @@ def check_resource_against_profile(resource, resource_type, specified_profile =
       end
 
       def resolve_path(elements, path)
+        return Inferno::FHIRPATH_EVALUATOR.evaluate(elements, path) if Inferno::FHIRPATH_EVALUATOR
```
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply lib/app/sequence_base.rb:746]: I'm not 100% sure. @czh-orz ?

File: lib/app/utils/basic_fhirpath_evaluator.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    private
+
+    def patch_path(path)
+      path = path.dup
+      path.sub!(/^[A-Z]\w*\./, '')
+      path.gsub!(/\bclass\b/, 'local_class')
+      path.gsub!(/\.where\(.*\)/, '')
+      as_type = path.scan(/\.as\((.*?)\)/).flatten.first
+      path.gsub!(/\.as\((.*?)\)/, capitalize_first_letter(as_type)) if as_type.present?
```
Comment (Jammjammjamm) [inline lib/app/utils/basic_fhirpath_evaluator.rb:26]:
```suggestion
      path.gsub!(/\.as\((.*?)\)/, as_type.upcase_first) if as_type.present?
```

**Review by Jammjammjamm — COMMENTED** [review]
This is looking pretty good. Still need to do some testing, which I won't be able to get to until tomorrow.

File: lib/app/utils/basic_fhirpath_evaluator.rb (hunk 5 above)
Comment (jason-crowley) [reply lib/app/utils/basic_fhirpath_evaluator.rb]: Replaced `capitalize_first_letter`

**Comment by Jammjammjamm** [comment]
I noticed this difference when running against our demo server:

dev:
Screen Shot 2020-09-04 at 9 21 16 AM

FI-930:
Screen Shot 2020-09-04 at 9 21 52 AM

**Comment by jason-crowley** [comment]
> I noticed this difference when running against our demo server:
> 
> dev:
> Screen Shot 2020-09-04 at 9 21 16 AM
> 
> FI-930:
> Screen Shot 2020-09-04 at 9 21 52 AM

Yes, I came across that issue as well. It seems to be non-deterministic which outcome happens, and I was able to reproduce that skip result on development by re-running the test multiple times.

**Review by Jammjammjamm — APPROVED** [review]

**Review by czh-orz — APPROVED** [review]
